# TASK 4.6. Combining & Exporting Data

## This script contains the following points:
1. Import data
2. Merge data
3. Export data

In [10]:
# Import libraries

import pandas as pd
import numpy as np
import os

# 01.Import data

In [11]:
# Set folder path
path = r'C:\Users\hachl\Downloads\4.3_orders_products'

In [12]:
# Import the 'orders_products_combined' dataframe from the pickle file

df_merged_large = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data', 'orders_products_combined.pkl'))

In [13]:
df_merged_large.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             0           0   2539329        1    prior             1   
2             0           0   2539329        1    prior             1   
3             0           0   2539329        1    prior             1   
4             0           0   2539329        1    prior             1   

   orders_day_of_week  order_hour_of_day  days_since_last_order  product_id  \
0                   2                  8                    NaN         196   
1                   2                  8                    NaN       14084   
2                   2                  8                    NaN       12427   
3                   2                  8                    NaN       26088   
4                   2                  8                    NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [14]:
# Check the shape of the imported dataframe

df_merged_large.shape

(32434489, 13)

### The previously merged data frame has the same number of rows and columns as when exported.

In [15]:
# Import the 'products_checked' csv file to merge it

df_prods = pd.read_csv(os.path.join(path,'02 Data','Prepared Data', 'products_checked.csv'), index_col = False)

In [16]:
# Check the output

df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [17]:
df_prods.shape

(49672, 6)

# 02.Merge data

###  I tried merging the two data frames on the common 'product_id' column but it gave me a ValueError: Cannot use name of an existing column for indicator column. I need to drop the '_merge' column as the argument 'indicator = True' is trying to create a column with the same name

In [18]:
# Drop unneccesary columns --> Error message due to '_merge' column already existing

df_merged_large = df_merged_large.drop(columns = ['_merge','eval_set','Unnamed: 0','Unnamed: 0.1'])

In [19]:
# Check the output

df_merged_large.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_last_order', 'product_id',
       'add_to_cart_order', 'reordered'],
      dtype='object')

In [20]:
# Drop unneccessary column from second dataframe

df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [21]:
# Check the output

df_prods.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices'], dtype='object')

### I retried running the merging command but this time it gave me a MemoryError: Unable to allocate 1.69 GiB for an array with shape (7, 32434489) and data type int64


## Dealing with Memory Errors: Plan A

- Avoid int64 data type because it consumes a lot of memory. Try int32 or int8.
- After every take run info() to see how much the memory usage decreased (in the bottom) but also head() because some of these changes may affect the data!

In [22]:
# Check memory usage of 'df_merged_large'
df_merged_large.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int64  
 1   user_id                int64  
 2   order_number           int64  
 3   orders_day_of_week     int64  
 4   order_hour_of_day      int64  
 5   days_since_last_order  float64
 6   product_id             int64  
 7   add_to_cart_order      int64  
 8   reordered              int64  
dtypes: float64(1), int64(8)
memory usage: 2.2 GB


In [23]:
# Check memory usage of products data set
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49672 entries, 0 to 49671
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49672 non-null  int64  
 1   product_name   49672 non-null  object 
 2   aisle_id       49672 non-null  int64  
 3   department_id  49672 non-null  int64  
 4   prices         49672 non-null  float64
dtypes: float64(1), int64(3), object(1)
memory usage: 1.9+ MB


In [24]:
# Change products data type from int64

df_prods['product_id']=df_prods['product_id'].astype('int32')
df_prods['aisle_id'] = df_prods['aisle_id'].astype('int32')
df_prods['department_id']=df_prods['department_id'].astype('int32')

In [25]:
# Check memory usage updated
df_prods.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49672 entries, 0 to 49671
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     49672 non-null  int32  
 1   product_name   49672 non-null  object 
 2   aisle_id       49672 non-null  int32  
 3   department_id  49672 non-null  int32  
 4   prices         49672 non-null  float64
dtypes: float64(1), int32(3), object(1)
memory usage: 1.3+ MB


In [26]:
# Check output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

### Attemped again a merge but memory error persists: MemoryError: Unable to allocate 1.93 GiB for an array with shape (8, 32434489) and data type int64

In [27]:
# Change data type for the other dataframe as well

df_merged_large['order_id']=df_merged_large['order_id'].astype('int32')
df_merged_large['user_id'] = df_merged_large['user_id'].astype('int32')
df_merged_large['order_number']=df_merged_large['order_number'].astype('int32')
df_merged_large['orders_day_of_week']=df_merged_large['orders_day_of_week'].astype('int8')
df_merged_large['order_hour_of_day']=df_merged_large['order_hour_of_day'].astype('int8')
df_merged_large['days_since_last_order']=df_merged_large['days_since_last_order'].astype('float16')
df_merged_large['product_id']=df_merged_large['product_id'].astype('int32')
df_merged_large['add_to_cart_order']=df_merged_large['add_to_cart_order'].astype('int32')
df_merged_large['reordered']=df_merged_large['reordered'].astype('int32')

In [29]:
# Check updated memory usage of 'df_merged_large'
df_merged_large.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 9 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   order_id               int32  
 1   user_id                int32  
 2   order_number           int32  
 3   orders_day_of_week     int8   
 4   order_hour_of_day      int8   
 5   days_since_last_order  float16
 6   product_id             int32  
 7   add_to_cart_order      int32  
 8   reordered              int32  
dtypes: float16(1), int32(6), int8(2)
memory usage: 866.1 MB


In [30]:
# Check output
df_merged_large.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  
0                    NaN         196                  1          0  
1                    NaN       14084                  2          0  
2                    NaN       12427                  3          0  
3                    NaN       26088                  4          0  
4                    NaN       26405                  5          0

In [31]:
# Try merging now that the memory usage decreased for both data sets

df_merged_final = df_merged_large.merge(df_prods, on = ['product_id'], indicator = True)

### Victory!!! 🎆🎆🎆🎆🎆🎆

In [32]:
# Confirm merge output through merge flag frequency --> result only in both, we know it´s a full-match independently.

df_merged_final['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [33]:
# Check output
df_merged_final.shape

(32404859, 14)

In [34]:
# Visualise the new data set
df_merged_final.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   15.0         196                  1          1   
2                   21.0         196                  1          1   
3                   29.0         196                  1          1   
4                   28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

# 03.Export data

In [35]:
# Export the new merged dataframe as a pickle file --> saves time and guarantees df integrity

df_merged_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'merged_data.pkl'))